modules/core/lib/Auth/Process/AttributeValueMap.php

```diff
@@ -0,0 +1,122 @@
+    */
+    private $replace = false;
+    
+    /**
+    * Initialize this filter.
+    *
+    * @param array $config  Configuration information about this filter.
+    * @param mixed $reserved  For future use.
+    */
+    public function __construct($config, $reserved) {
+        parent::__construct($config, $reserved);
+
+        assert('is_array($config)');
+
+        /* Validate configuration. */
+        foreach ($config as $name => $value) {
+            if (is_int($name)) {
+                // check if this is an option
+                if ($value === '%replace') {
+                        $this->replace = true;
+                } elseif ($value === '%keep') {
+                        $this->keep = true;
+                } else {
+                        throw new SimpleSAML_Error_Exception('Unknown flag : ' . var_export($value, true));
+                }
+                continue;
+            }
+
+            // Set targetattribute
+            if ($name === 'targetattribute') {
+                $this->targetattribute = $value;
+            }
+
+            // Set sourceattribute
+            if ($name === 'sourceattribute') {
+                $this->sourceattribute = $value;
+            }
+        
+            // Set values
+            if ($name === 'values') {
+                $this->values = $value;
+            }
+        }
+    }
+
+
+    /**
+        * Apply filter to add groups attribute.
+        *
+        * @param array &$request  The current request
+        */
+    public function process(&$request) {
+        SimpleSAML_Logger::debug('AttributeValueMap - process');
+
+        assert('is_array($request)');
+        assert('array_key_exists("Attributes", $request)');
+        $attributes =& $request['Attributes'];
+
+        // Make sure sourceattribute exists
+        assert('array_key_exists($this->sourceattribute, $attributes)');
+        // Make sure the targetattribute is set
+        assert('is_string($this->targetattribute)');
+        
+        $sourceattribute = $attributes[$this->sourceattribute];
+        $targetvalues = array();
+
+        if (is_array($sourceattribute)) {
+            foreach ($this->values as $value => $require) {
+                if (count(array_intersect($require, $sourceattribute)) > 0) {
+                    SimpleSAML_Logger::debug('AttributeValueMap - intersect match for ' . $value);
+                    $targetvalues[] = $value;
+                }
+            }
+        }
+
+        if (count($targetvalues) > 0) {
+            if ($this->replace or !@is_array($attributes[$this->targetattribute])) {
+                $attributes[$this->targetattribute] = $targetvalues;
+            } else {
+                $attributes[$this->targetattribute] = array_unique(array_merge($attributes[$this->targetattribute], $targetvalues));
+            }
+        }
+
+        if (!$this->keep) {
+            unset($attributes[$this->sourceattribute]);
+        }
+    }
+}
+
+?>
```

Comment:
Leaving the closing `?>` makes it possible to add contents after it, even some that could go unnoticed, like newlines, spaces or tabs. Those characters will be printed to standard output, forcing PHP to send the headers. And once the headers have been sent, you can't send more headers.

So if something is printed by the end of a file and you want to make a redirection after importing that file, you won't be able. Imagine you have a trailing newline in a processing filter like this, and the result of the filter should be to redirect to another page. It won't work because headers were sent when you included the file.

Same applies to setting cookies and anything else that needs to send an HTTP header to the browser.